### PR TITLE
Add "Internal" flag to Metric::Flags to allow the ForEach loop to ignore or skip metrics with that flag set

### PR DESCRIPTION
**Pull request by DiazAlan**
Creates a special flag "hidden" to use in Metrics. This will allow the ForEach Loops for different metrics to check for this flag and skip flags that have this set. Currently only Gauge has this implemented to help close this PR [23921](https://github.com/envoyproxy/envoy/pull/23921).

### CODE REVIEW

**Comment by DiazAlan** [comment]
@jmarantz @yurykats @pradeepcrao @stevenzzzz 
I haven't added tests or support for hot restart but that's coming soon. I appreciate any comments or suggestions in the meantime.

File: envoy/stats/stats.h
Comment (jmarantz) [inline envoy/stats/stats.h:96]: We should say more about exactly what APIs this affects. I was thinking maybe we *would* want to hit this in forEach loops but we would want to leave it up to the stats sink and admin panel to skip them.

Note that for stats-sinks there's already a sink-predicate concept this could be an extension of that.

I'm wondering also if it might be interesting for our own devs relying on this feature to be able to look at the hidden stats from the admin panel, say with a `/stats?hidden` parameter.

Comment (jmarantz) [reply envoy/stats/stats.h:96]: I was thinking it's cleaner to have forEach hit all the data structures cleanly; the caller supplying the functor can then apply used() and hidden() (I slightly prefer 'hidden' to 'internal') per policy they choose.

Comment (yurykats) [reply envoy/stats/stats.h:96]: That sounds reasonable to me as well, especially if there is precedent. It would likely also mean that we won't need to do anything special for the hot restart case!

Comment (DiazAlan) [reply envoy/stats/stats.h:96]: I reverted the change in the ForEach loops

File: source/common/stats/thread_local_store.cc
Comment (jmarantz) [inline source/common/stats/thread_local_store.cc:598]: a minor point but we should leave a TODO to rename ImportMode go be something more generic. Not sure what name I like best but it's not just import that's being controlled anymore.

Maybe just Mode (which is already inside Gauge namespace) so it'll read like Gauge::Mode::Accumulate or Gauge::Mode::NeverImport or Gauge::Mode::Hidden, which seems clear enough and not redundant or verbose.

Comment (jmarantz) [reply source/common/stats/thread_local_store.cc:598]: I am not asking you to do it in this PR -- I think big refactors touching hundreds of lines of unrelated files with a simple rename are easier to do as a separate PR that changes zero semantics

Comment (DiazAlan) [reply source/common/stats/thread_local_store.cc]: Done. I added a few TODOs for myself at different spots

File: envoy/stats/stats.h
Comment (jmarantz) [inline envoy/stats/stats.h:104]: It occurs to me that, as this applies only to gauges, we could encode the 4 possible states into 2 bits rather than using 4 bits.
```
00 = Uninitialized
01 = Accumulate
10 = NeverImport
11 = Hidden
```
or something like that. I'm not that concerned about saving the 2 flag bits but it's nicer not to have illegal states we should arguably check for.

I wanted to share this thought, but leaving the code as you have it is likely simpler than encoding as I suggested. But could you leave a note here that we are encoding 4 states in 4 bits and could easily collapse to 2 if we ever ran low?

**Review by jmarantz — COMMENTED** [review]
there will be follow-up PRs to skip hidden stats in admin `/stats`, unless user has specified `?hidden` right?

File: test/common/stats/allocator_impl_test.cc
Comment (DiazAlan) [reply test/common/stats/allocator_impl_test.cc:262]: I know by default it doesn't care about hidden. In this test it does pass a functor that tells the foreach loop to ignore the Hidden Gauge. Does that suffice for the test?

Comment (jmarantz) [reply test/common/stats/allocator_impl_test.cc:262]: It's not that it doesn't suffice; it's just got a lot of irrelevant stuff in it :)

File: source/common/stats/allocator_impl.cc
Comment (jmarantz) [reply source/common/stats/allocator_impl.cc:222]: Yeah that's right; popcount would not be appropriate. We could do
```
  ASSERT(flags_ & Flags::NeverImport + flags_ & Flags::LogicAccumulate + flags_ & Flags::LogicHidden <= 1);
```

File: envoy/stats/stats.h
Comment (DiazAlan) [reply envoy/stats/stats.h:104]: As per another comment, since a gauge can be both used and have different import modes. This would mean that it would have more than just 4 "states" so it can't be represented by 2 bits. And I worry that making it 3 bits to represent 8 states would start to make things a little complicated to read.

Comment (jmarantz) [reply envoy/stats/stats.h:104]: The used-bit can stay as it is.

I was just talking about blocking out 2 bits for the 4 gauge-mode related states. It would be a little messy to code which is why I'm suggesting just a comment and an assert.

File: source/common/stats/allocator_impl.cc
Comment (DiazAlan) [reply source/common/stats/allocator_impl.cc:222]: Done. I added some ASSERTs in different spots

File: source/common/stats/thread_local_store.cc
Comment (DiazAlan) [reply source/common/stats/thread_local_store.cc:600]: I added two tests to thread_local_store_test to test rejectAll and reject by Regex

File: test/common/stats/allocator_impl_test.cc
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:143]: GaugeSharedPtr (exact type is not evident from context).

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1153]: Why is this needed? The above test in the same test-fixture does not. I think this test might not have any mocks.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1156]: Why such a complex regex? Why not simply '".*"'?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1164]: I think you can replace one of these with a case where the ImportMode is something other than hidden, and have the test validate that hte returned gauge is the null gauge.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1168]: not needed, I think.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1177]: again I think you can replace one of these with an import-mode of something other than Hidden, so that you can verify that the rejection is working in that case, and make it clearer that Hidden is treated specially.

**Review by jmarantz — COMMENTED** [review]
@pradeepcrao I lost track of what in this PR depends on your change. It's moot now, but I thought your change just optimized the Accept path which we are not testing.

File: envoy/stats/stats.h
Comment (jmarantz) [inline envoy/stats/stats.h:138]: I think we should specify here what 'hidden' means. It's fine to give the use-case also. But we should specify:
 * will be transferred on hot-restart
 * will not show up in stats-sinks and admin. Though as I've mentioned it might be helpful when reasoning about behavior to make /stats?hidden work -- maybe calling it that the stats endpoint should *only* show hidden stats and nothing else when enabled.

Speaking of which, are we going to include the 'hidden' behavior described here in this PR? I think it would probably make sense. Right now the only effect of 'hidden' is to block disabling of those stats.

**Review by jmarantz — COMMENTED** [review]
per IM I think I need to alter an earlier suggestion:

1. I think we should call the new enum HiddenAccumulate
2. That should set both the 'hidden' and 'accumulate' bits in the flags.
3. We should obviously *not* then have asserts that prevent setting multiple bits (though Accumulate and NeverImport would be mutually exclusive as they are now).
4. We could in principle also have a HiddenNeverImport option in the enum, but we don't need it now so I suggest we not bother adding it.

/wait

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Done

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Removed

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Fair enough. Changed

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Done

File: envoy/stats/stats.h
Comment (DiazAlan) [reply envoy/stats/stats.h]: I was thinking the hidden behavior would be included in a follow up PR

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [inline test/common/stats/thread_local_store_test.cc:1153]: Could you please also add a test that with a stat matcher that uses an `inclusion_list` ?

Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:1153]: Actually you'll also need tests with prefix matches with both inclusion and exclusion lists (to prevent regressions in the future).

Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc:1182]: Done

**Comment by pradeepcrao** [comment]
> @pradeepcrao I lost track of what in this PR depends on your change. It's moot now, but I thought your change just optimized the Accept path which we are not testing.

This PR requires that we short circuit `slowRejects` if fast_result is `FastResult::Matches`

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [inline test/common/stats/thread_local_store_test.cc:1155]: s/cluster/cluster./

File: test/common/stats/thread_local_store_test.cc
Comment (pradeepcrao) [inline test/common/stats/thread_local_store_test.cc:1170]: s/cluster/cluster./

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Done

File: test/common/stats/thread_local_store_test.cc
Comment (DiazAlan) [reply test/common/stats/thread_local_store_test.cc]: Done

File: envoy/stats/stats.h
Comment (jmarantz) [inline envoy/stats/stats.h:96]: End sentences with a period.

File: test/common/stats/allocator_impl_test.cc
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:145]: add `EXPECT_TRUE(hidden_gauge->hidden());`

File: test/common/stats/allocator_impl_test.cc
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:148]: add `EXPECT_FALSE(non_hidden_gauge->hidden());` and add a third case for Gauge::ImportMode::NeverImport to check it's also not hidden.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [reply test/common/stats/thread_local_store_test.cc:1196]: it's still complex.

**Review by jmarantz — COMMENTED** [review]
This basically looks great; a few nits remain.

File: source/common/stats/thread_local_store.cc
Comment (jmarantz) [inline source/common/stats/thread_local_store.cc:588]: we'll need to remember to adjust this if we ever add HiddenNeverImport. No action required on this comment now, but just putting this out there :)

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [reply test/common/stats/thread_local_store_test.cc:1196]: When I said "it's still complex", github was showing me a different version. I still haven't totally wrapped my arms around the data model for github reveiws.

This is fine. It's a little subtle that you are intending to create an inclusion list requiring a single upper-case letter, if I'm reading your intent right. Maybe just add a comment?

**Review by jmarantz — COMMENTED** [review]
/wait

File: test/common/stats/allocator_impl_test.cc
Comment (DiazAlan) [reply test/common/stats/allocator_impl_test.cc:149]: Done

File: test/common/stats/allocator_impl_test.cc
Comment (DiazAlan) [reply test/common/stats/allocator_impl_test.cc:145]: Done

File: envoy/stats/stats.h
Comment (DiazAlan) [reply envoy/stats/stats.h]: Done

File: envoy/stats/stats.h
Comment (DiazAlan) [reply envoy/stats/stats.h:144]: Added a TODO above to add the admin/stats-sink functionality

File: envoy/stats/stats.h
Comment (pradeepcrao) [inline envoy/stats/stats.h:96]: Please rephrase as to not mention the flag. That's an implementation detail and we're explicitly ignoring the flag value for counters, text readouts and histograms.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1167]: nit: drop blank link here.

File: source/common/stats/thread_local_store.cc
Comment (pradeepcrao) [inline source/common/stats/thread_local_store.cc:598]: Could you please add a comment here stating that we will always create hidden stats and ignore the stats_matcher unless we're rejecting all stats?

**Review by jmarantz — APPROVED** [review]
lgtm!

@pradeepcrao do you want to finish your review before we hand off to Greg?

**Review by jmarantz — APPROVED** [review]
lgtm!

@pradeepcrao do you want to finish your review before we hand off to Greg?

**Review by jmarantz — DISMISSED** [review]
lgtm!

@pradeepcrao do you want to finish your review before we hand off to Greg?

**Review by pradeepcrao — COMMENTED** [review]
LGTM. (A few nits regarding comments).

File: source/common/stats/thread_local_store.cc
Comment (DiazAlan) [reply source/common/stats/thread_local_store.cc:598]: It's currently implemented to ignore rejectAll if the stat is hidden. Is this what we would want considering deferred stats needs the hidden gauges?

**Comment by jmarantz** [comment]
@ggreenway for senior maintainer approval.

**Pull request merged**
